### PR TITLE
DataSource: Allow data source plugins to set query default values

### PR DESCRIPTION
**Pull request by sunker**
**What this PR does / why we need it**:
It would be good if the datasource api provided a way for a datasource to set a default query that is applied whenever new query rows are added. That way we could get rid of boilerplate code in the editor and also prevent unnecessary rerenders that are triggered when a data source call `onChange` after setting default values. Curious to hear your thoughts on this approach.

### CODE REVIEW

**Comment by sunker** [comment]
@grafana/plugins-platform-frontend can you take a look at this?

File: packages/grafana-data/src/types/datasource.ts [hunk 1]
```diff
@@ -216,6 +216,7 @@ abstract class DataSourceApi<
     this.type = instanceSettings.type;
     this.meta = instanceSettings.meta;
     this.uid = instanceSettings.uid;
+    this.defaultQuery = {};
```
Comment (torkelo) [inline packages/grafana-data/src/types/datasource.ts:219]: wonder if this should instead be a function, so we can have more context specific default queries..

getDefaultQuery(app: CoreApp): Partial<TQuery> {
} 

That we can more easily support different defaults for Explore or an Alerting query.

**Review by torkelo — COMMENTED** [review]
Think this looks like a good suggestion. Added one suggestion to make it more future proof / flexible

**Comment by sunker** [comment]
> Think this looks like a good suggestion. Added one suggestion to make it more future proof / flexible

Yes good idea! Feedback applied. Can you take a second look?

File: public/app/features/explore/state/query.ts [hunk 2]
```diff
@@ -213,10 +214,17 @@ export const clearCacheAction = createAction<ClearCachePayload>('explore/clearCa
...
+export function addQueryRow(
+  exploreId: ExploreId,
+  index: number,
+  datasource: DataSourceApi | undefined | null
+): ThunkResult<void> {
   return (dispatch, getState) => {
     const queries = getState().explore[exploreId]!.queries;
-    const query = generateEmptyQuery(queries, index);
+    const query = {
+      ...(datasource?.getDefaultQuery && datasource.getDefaultQuery(CoreApp.Explore)),
```
Comment (Elfo404) [inline public/app/features/explore/state/query.ts:225]: just a nit:
```suggestion
      ...datasource?.getDefaultQuery?.(CoreApp.Explore),
```

File: public/app/features/query/components/QueryGroup.tsx [hunk 3]
```diff
@@ -137,6 +138,7 @@ export class QueryGroup extends PureComponent<Props, State> {
     const ds = !dsSettings?.meta.mixed ? dsSettings : defaultDataSource;
 
     return {
+      ...(this.state.dataSource?.getDefaultQuery && this.state.dataSource.getDefaultQuery(CoreApp.PanelEditor)),
```
Comment (Elfo404) [inline public/app/features/query/components/QueryGroup.tsx:141]: just a nit:
```suggestion
      ...this.state.dataSource?.getDefaultQuery?.(CoreApp.PanelEditor),
```

**Review by Elfo404 — APPROVED** [review]
LGTM and totally makes sense. left just a nit

**Review by torkelo — APPROVED** [review]

**Comment by aocenas** [comment]
Would it maybe make sense to add more context information? For example would it make sense to know panel type (if in dashboard)? So that we can have appropriate default based on panel type which may need different queries?

**Comment by aocenas** [comment]
> Also can we make it https://github.com/internal or something like that for now so we try it in our data sources first?

> Don't think that's technically feasible? Or do you mean for core data sources only? If so I rather not because I'd like to use this in our external plugins asap.

So not sure how it works but I seen some tags like that basically saying that this API should not be use by third party developers yet or just on their own risk. So that we can easier change it if needed and not break too many plugins.

**Comment by sunker** [comment]
> So not sure how it works but I seen some tags like that basically saying that this API should not be use by third party developers yet or just on their own risk. So that we can easier change it if needed and not break too many plugins.

@aocenas  maybe tag the method as @alpha?

**Pull request merged**